nil pointer deference when updating process group during deploy

https://github.com/superfly/flyctl/blob/41691d428650e68060b159d64e0d24e99fbeca6c/internal/command/deploy/plan.go#L340

panic: runtime error: invalid memory address or nil pointer dereference
2024-11-22T14:04:49.7325598Z [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x20f983e]
2024-11-22T14:04:49.7326218Z
2024-11-22T14:04:49.7326919Z goroutine 73418 [running]:
2024-11-22T14:04:49.7327955Z github.com/superfly/flyctl/internal/command/deploy.(*machineDeployment).updateProcessGroup.func1()
2024-11-22T14:04:49.7329492Z    /home/runner/work/flyctl/flyctl/internal/command/deploy/plan.go:340 +0x1e
2024-11-22T14:04:49.7330440Z golang.org/x/sync/errgroup.(*Group).Go.func1()
2024-11-22T14:04:49.7331597Z    /home/runner/go/pkg/mod/golang.org/x/sync@v0.9.0/errgroup/errgroup.go:78 +0x56
2024-11-22T14:04:49.7332605Z created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 68623
2024-11-22T14:04:49.7333837Z    /home/runner/go/pkg/mod/golang.org/x/sync@v0.9.0/errgroup/errgroup.go:75 +0x96
2024-11-22T14:04:49.7446409Z ##[error]Process completed with exit code 2.

--- a/internal/command/deploy/plan.go
+++ b/internal/command/deploy/plan.go
@@ -387,7 +387,7 @@ func (md *machineDeployment) updateProcessGroup(ctx context.Context, machineTupl
 				sl.LogStatus(statuslogger.StatusFailure, err.Error())
 				span.RecordError(err)
 
-				return fmt.Errorf("failed to update machine %s: %w", oldMachine.ID, err)
+				return fmt.Errorf("failed to update machine %s: %w", machineID, err)
 			}
 
 			return nil

--- a/internal/command/deploy/plan_test.go
+++ b/internal/command/deploy/plan_test.go
@@ -520,6 +520,99 @@ func TestUpdateMachinesWithNewMachine(t *testing.T) {
 	assert.NoError(t, err)
 }
 
+// TestUpdateProcessGroupScaleUpErrorDoesNotPanic verifies that when
+// updateMachineWChecks returns an error for a scale-up pairing
+// (oldMachine=nil), updateProcessGroup formats the error without
+// dereferencing oldMachine. Before the fix, this panicked inside an
+// errgroup goroutine and crashed the process.
+func TestUpdateProcessGroupScaleUpErrorDoesNotPanic(t *testing.T) {
+	t.Parallel()
+
+	ctx := withQuietIOStreams(context.Background())
+
+	// Empty old state and a single new machine produces exactly one pairing
+	// {oldMachine: nil, newMachine: ...}, hitting the scale-up code path.
+	oldMachines := []*fly.Machine{}
+	newMachines := []*fly.Machine{
+		{
+			ID:         "new-machine",
+			State:      "started",
+			Region:     "iad",
+			HostStatus: fly.HostStatusOk,
+			Config: &fly.MachineConfig{
+				Image: "image1",
+			},
+		},
+	}
+
+	flapsClient := &mock.FlapsClient{
+		AcquireLeaseFunc: func(ctx context.Context, appName, machineID string, ttl *int) (*fly.MachineLease, error) {
+			return &fly.MachineLease{
+				Data: &fly.MachineLeaseData{
+					Nonce: machineID + "nonce",
+				},
+			}, nil
+		},
+		ReleaseLeaseFunc: func(ctx context.Context, appName, machineID, nonce string) error {
+			return nil
+		},
+		RefreshLeaseFunc: func(ctx context.Context, appName, machineID string, ttl *int, nonce string) (*fly.MachineLease, error) {
+			return &fly.MachineLease{
+				Status: "success",
+				Data:   &fly.MachineLeaseData{Nonce: nonce},
+			}, nil
+		},
+		// LaunchFunc fails so updateOrCreateMachine's scale-up arm returns an
+		// error, which propagates back through updateMachineWChecks to the
+		// error-formatting branch in updateProcessGroup.
+		LaunchFunc: func(ctx context.Context, appName string, builder fly.LaunchMachineInput) (out *fly.Machine, err error) {
+			return nil, assert.AnError
+		},
+		ListFunc: func(ctx context.Context, appName, state string) ([]*fly.Machine, error) {
+			return oldMachines, nil
+		},
+		GetFunc: func(ctx context.Context, appName, machineID string) (*fly.Machine, error) {
+			return &fly.Machine{
+				ID:         machineID,
+				State:      "started",
+				HostStatus: fly.HostStatusOk,
+				Config:     &fly.MachineConfig{Image: "image1"},
+			}, nil
+		},
+	}
+
+	ctx = flapsutil.NewContextWithClient(ctx, flapsClient)
+	md := &machineDeployment{
+		flapsClient: flapsClient,
+		io:          iostreams.FromContext(ctx),
+		app: &flaps.App{
+			Name: "myapp",
+		},
+		appConfig:       &appconfig.Config{AppName: "myapp"},
+		waitTimeout:     10 * time.Second,
+		deployRetries:   0,
+		maxUnavailable:  3,
+		skipSmokeChecks: false,
+	}
+
+	oldAppState := &AppState{Machines: oldMachines}
+	newAppState := &AppState{Machines: newMachines}
+
+	// pushForward=false makes the error propagate immediately instead of
+	// retrying, so we can assert on it.
+	settings := updateMachineSettings{
+		pushForward:          false,
+		skipHealthChecks:     false,
+		skipSmokeChecks:      false,
+		skipLeaseAcquisition: false,
+	}
+
+	err := md.updateMachinesWRecovery(ctx, oldAppState, newAppState, nil, settings)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "new-machine",
+		"error must reference the new machine's ID; if oldMachine.ID was used, the test would have panicked instead")
+}
+
 func TestUpdateOrCreateMachine(t *testing.T) {
 	t.Parallel()
 

